ministryofjustice/correspondence_tool_staff
# CT-3034, CT-3036 Fixed the bugs related to offender sar cases

## Description

- Minor fixes for close/open reports like column name, order, and wrong logic
- Bug fix for performance report:  if the month is not within the range of the performance report, it will throw an error, 
- Changed the Redis.exists() to Redis.exists? due to upgrade of gem version 
- Added translation for even of "destroy"

## Self-review checklist


### Screenshots

- Minor fixes for close/open reports like column name, order, and wrong logic
For open report
before 
<img width="1062" alt="open_case_before" src="https://user-images.githubusercontent.com/26111671/91472685-bb62c080-e88f-11ea-8b6a-6127a53e0bfd.png">
after 
<img width="1006" alt="Screenshot 2020-08-27 at 18 22 33" src="https://user-images.githubusercontent.com/26111671/91474696-78561c80-e892-11ea-91bd-83d5e64ad129.png">

For close report
before 
<img width="893" alt="close_case_before" src="https://user-images.githubusercontent.com/26111671/91472734-ccabcd00-e88f-11ea-9f6d-67f9330227e4.png">
After 
<img width="1157" alt="close_case_after_fix" src="https://user-images.githubusercontent.com/26111671/91472746-d0d7ea80-e88f-11ea-96f6-b124ec05e825.png">


- Bug fix for performance report:  if the month is not within the range of the performance report, it will throw an error, 
This has been covered by new test spec I added in this PR

- Changed the Redis.exists() to Redis.exists? due to upgrade of gem version 
Fixed the error message 
```
Redis#exists(key)` will return an Integer in redis-rb 4.3. `exists?` returns a boolean, you should use it instead. To opt-in to the new behavior now you can set Redis.exists_returns_integer = true. To disable this message and keep the current (boolean) behaviour of 'exists' you can set `Redis.exists_returns_integer = false`, but this option will be removed in 5.0. (/usr/src/app/app/services/stats/base_monthly_performance_report.rb:94:in `block in report_details'
````
- Added translation for even of "destroy"
Before
<img width="806" alt="destroy_before_add" src="https://user-images.githubusercontent.com/26111671/91473314-9fabea00-e890-11ea-9c58-dfb700848a65.png">
After
<img width="823" alt="destroy_after_add" src="https://user-images.githubusercontent.com/26111671/91473327-a33f7100-e890-11ea-8cba-0f01ab55c83d.png">


### Related JIRA tickets

https://dsdmoj.atlassian.net/secure/RapidBoard.jspa?rapidView=25&modal=detail&selectedIssue=CT-3036
https://dsdmoj.atlassian.net/secure/RapidBoard.jspa?rapidView=25&modal=detail&selectedIssue=CT-3034
### Deployment

If you try to run this branch from local, please make sure you have run sidekiq locally as well
### Manual testing instructions

Download the open case report
1 - Login as Branston user
2 - click "All open cases" table
3 - Click the "Download cases" link at right top of the case list
4 - A csv file should be downloadable, and check the columns and vlaues
Download the close cases report
1 - Login as Branson usesr
2 - Choose the "Reports" menu from top
3 - Click "Create report" button 
4 - Choose "Offender SAR" type
5 - Choose sub report "Closed case report"
6 - Type in the time period you want to produce 
7 - a Message should be popped from top which give the link for you to download it
8 - a zip file will be downloaded after you click that link
9 - unzip it and check the columns and values from the csv file 
Download the performance report
1 - repeat 1 to 4 as above
2- Choose sub report "Monthly report (Offender SARs)"
3 - a Message should be popped from top which give the link for you to download it
if the message tells you this report is not ready yet please wait for a few seconds to try again
4 - a excel file should be downloaded
